src/main/java/io/cdap/plugin/gcp/spanner/sink/SpannerOutputFormat.java

```diff
@@ -231,4 +243,125 @@ public void abortTask(TaskAttemptContext taskAttemptContext) {
+    Schema unionSchema = componentSchema.getUnionSchema(0);
+    if (unionSchema == null) {
+      throw new IllegalStateException("Union schema of '" + componentSchema.getRecordName() + "' is null");
+    }
+
+    Schema.LogicalType logicalType = unionSchema.getLogicalType();
+    if (logicalType != null) {
+      switch (logicalType) {
+        case DATE:
+          Collection<LocalDate> dateList = toCollection(name, logicalType.toString(), values);
+          builder.set(name)
+            .toDateArray(
+              dateList.stream()
+                .map(date -> Date.fromYearMonthDay(date.getYear(), date.getMonthValue(), date.getDayOfMonth()))
+                .collect(Collectors.toList())
+            );
+          break;
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          Collection<ZonedDateTime> timestampList = toCollection(name, logicalType.toString(), values);
+          builder.set(name)
+            .toTimestampArray(
+              timestampList.stream()
+                .map(ts -> Timestamp.ofTimeSecondsAndNanos(ts.toEpochSecond(), ts.getNano()))
+                .collect(Collectors.toList())
+            );
+          break;
+        default:
+          throw new IllegalStateException("Array of '" + logicalType + "' logical type currently not supported.");
+      }
+    } else {
+      Schema.Type type = unionSchema.getType();
+      switch (type) {
+        case BOOLEAN:
+          Collection<Boolean> listBoolean = toCollection(name, type.toString(), values);
+          builder.set(name).toBoolArray(listBoolean);
+          break;
+        case STRING:
+          Collection<String> listString = toCollection(name, type.toString(), values);
+          builder.set(name).toStringArray(listString);
+          break;
+        case LONG:
+          Collection<Long> listValue = toCollection(name, type.toString(), values);
+          builder.set(name).toInt64Array(listValue);
+          break;
+        case DOUBLE:
+          Collection<Double> listDouble = toCollection(name, type.toString(), values);
+          builder.set(name).toFloat64Array(listDouble);
+          break;
+        case INT:
+          Collection<Integer> integerList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toInt64Array(
+              integerList.stream()
+                .map(Integer::longValue)
+                .collect(Collectors.toList())
+            );
+          break;
+        case FLOAT:
+          Collection<Float> floatList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toFloat64Array(
+              floatList.stream()
+                .map(Float::doubleValue)
+                .collect(Collectors.toList())
+            );
+          break;
+        case BYTES:
+          Collection<byte[]> bytesList = toCollection(name, type.toString(), values);
```

Comment:
CDAP bytes can be either a byte[] or a ByteBuffer. If it's a ByteBuffer it needs to be converted somewhere, either in toCollection or in the stream below

Reply:
Changed to use transformation from realization for BYTES type, made it similar to Datastore plugin and moved to RecordToEntityTransformer class.